### PR TITLE
Both part1 and part2 completed

### PR DESCRIPTION
**Pull request by JonathanYK**

### CODE REVIEW

File: pulu/awsSecrets.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"username":"postgres","password":"postgres","engine":"postgres","host":"postgres-instanceabae9b4.cck9ljnvsu6v.us-east-1.rds.amazonaws.com","port":5432,"dbname":"canddatabase","dbInstanceIdentifier":"postgres-instanceabae9b4"}
```
Comment (JonathanYK) [inline pulu/awsSecrets.json]: left that here for a reason - illustrate the file with demo secret attributes.

File: .vscode/launch.json
Comment (seniakalma) [inline .vscode/launch.json:1]: I don't think this file is needed

File: .vscode/settings.json
Comment (seniakalma) [inline .vscode/settings.json:1]: I don't think this file is needed

File: pulu/AWS_Infra_Visualization.png
Comment (seniakalma) [inline pulu/AWS_Infra_Visualization.png:1]: Awesome!

File: pulu/awsSecrets.json (hunk 1 above)
Comment (seniakalma) [reply pulu/awsSecrets.json]: V

File: pulu/dev/index.ts [hunk 2]
```diff
@@ -0,0 +1,541 @@
+import * as pulumi from "@pulumi/pulumi"
+import * as aws from "@pulumi/aws";
+import * as awsx from "@pulumi/awsx";
+import * as docker from "@pulumi/docker";
+import * as childProcess from 'child_process';
+import * as fs from 'fs';
+
+const region: string = "us-east-1"
+
+const availabilityZoneA: string = "us-east-1a";
```
Comment (seniakalma) [inline pulu/dev/index.ts:10]: Maybe just save it as 'a' and 'b' and append it to the region. Will make changing a region less broke prone

File: pulu/dev/index.ts [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+const availabilityZoneB: string = "us-east-1b";
+
+const availabilityZoneRDSc: string = "us-east-1c";
+const availabilityZoneRDSd: string = "us-east-1d";
+
+
+// pull aws secrets
+const filePath = "../awsSecrets.json";
```
Comment (seniakalma) [inline pulu/dev/index.ts:18]: Part of a config? First line of the file?

File: pulu/dev/index.ts [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+  vpcId: candVpc.id,
+  tags: {
+    Name: "cand-vpc-gw"
+  },
+});
+export const vpcGwId = vpcGw.id;
+
+
+// candidate main security group
+const candSg = new aws.ec2.SecurityGroup("cand-security-group", {
```
Comment (seniakalma) [inline pulu/dev/index.ts:48]: Awesome

File: pulu/awsSecrets.json (hunk 1 above)
Comment (seniakalma) [reply pulu/awsSecrets.json]: You can write this comment in the file ("stub values") or just leave them empty

File: pulu/dev/index.ts [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+  
+  vpcId: candVpc.id,
+  tags: {
+    Name: "cand-targetgroup-8085"
+  },
+});
+export const candTGId = candTG.id;
+
+
+// listener for port 8085
```
Comment (seniakalma) [inline pulu/dev/index.ts:383]: A lot of usage for port 8085, maybe you can change it to a 'appPort' variable

File: pulu/dev/index.ts [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+              "awslogs-group": candLogGroup.name,
+              "awslogs-stream-prefix": "--candImage Logs--"
+          }
+        },
+        cpu: 256,
+        memory: 128,
+        environment: [
+          {
+            name: "DB_HOST",
+            value: pgInstance.endpoint,
```
Comment (seniakalma) [inline pulu/dev/index.ts:447]: Looks great!

File: pulu/dev/index.ts [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+  },
+}, fargateServiceOptions);
+export const candFargateServiceId = candFargateService.service.id;
+const candFargateServiceName = candFargateService.service.apply(service => service.name);
+
+
+// create Dashboard at the end:
+const candDashboard = pulumi.all([]).apply(([]) => {
+
+  // CloudWatch Dashboard with CPU and memory Utilization
```
Comment (seniakalma) [inline pulu/dev/index.ts:484]: I like it

File: pulu/initUsersTable.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
+import sys
```
Comment (seniakalma) [inline pulu/initUsersTable.py]: As it's a standalone file maybe a 2 line explanation?

File: pulu/initUsersTable.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    cursor.execute(createUsersTableQuery)
+
+    # Commit the transaction to save the changes
+    connection.commit()
+
+except psycopg2.errors.DuplicateTable:
+    print(f"""Table "{create_table_name}" already exists.""")
+
+except (Exception, psycopg2.Error) as error:
+    print(f"Error while connecting to PostgreSQL on RDS: {error}")
```
Comment (seniakalma) [inline pulu/initUsersTable.py:53]: Nice

File: pulu/prod/index.ts [hunk 10]
```diff
@@ -0,0 +1,549 @@
+import * as pulumi from "@pulumi/pulumi"
+import * as aws from "@pulumi/aws";
+import * as awsx from "@pulumi/awsx";
+import * as docker from "@pulumi/docker";
+import * as childProcess from 'child_process';
+import * as fs from 'fs';
+
+const region: string = "us-west-2"
```
Comment (seniakalma) [inline pulu/prod/index.ts:8]: Same comment (region+a/b/c/d)

File: pulu/retriveAwsSecrets.js
Comment (seniakalma) [inline pulu/retriveAwsSecrets.js:1]: Short exp

File: pulu/retriveAwsSecrets.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+                    }
+                    else {
+                        console.log("Retrived AWS secrets successfully!");
+                        return [2 /*return*/, secretJson];
+                    }
+                    return [3 /*break*/, 3];
+                case 2:
+                    error_1 = _b.sent();
+                    throw error_1;
+                case 3: return [2 /*return*/];
```
Comment (seniakalma) [inline pulu/retriveAwsSecrets.js:74]: /*return*/

File: pulu/retriveAwsSecrets.ts
Comment (seniakalma) [inline pulu/retriveAwsSecrets.ts:1]: Short exp

File: serv/.env [hunk 12]
```diff
@@ -0,0 +1,10 @@
+HOST="localhost"
+PORT="8085"
+
+# All DB related parameters must be defined in AWS secrets manager:
```
Comment (seniakalma) [inline serv/.env:4]: Remove

File: serv/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,13 @@
+FROM node:lts-alpine as candidates
```
Comment (seniakalma) [inline serv/Dockerfile:1]: Why is the 'as candidates' needed?

File: serv/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  web:
+    build:
+      context: ./
+      target: candidates
+    volumes:
+      - .:/src
+    command: npm run dev
+    ports:
+      - "3001:3001"
+    #depends_on:
```
Comment (seniakalma) [inline serv/docker-compose.yml]: Remove?

File: serv/src/db/init.sql [hunk 15]
```diff
@@ -0,0 +1,11 @@
+CREATE USER compose;
+CREATE DATABASE apitests;
+GRANT ALL PRIVILEGES ON DATABASE apitests TO compose;
```
Comment (seniakalma) [inline serv/src/db/init.sql:3]: What's the meaning of this?

File: serv/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
+import * as dotenv from "dotenv";
+import express from "express";
+import cors from "cors";
+import { router } from "./candidates/endpoints.router";
+
+dotenv.config();
+
+envVarValidation();
+
+const HOST = process.env.HOST;
```
Comment (seniakalma) [inline serv/src/index.ts:10]: Throw error on the missing of those?

File: serv/tsconfig.json
Comment (seniakalma) [inline serv/tsconfig.json:1]: What's the usage of this?

**Review by seniakalma — COMMENTED** [review]
Looks very very good. Goob job. Small comments.

File: .vscode/launch.json
Comment (JonathanYK) [reply .vscode/launch.json:1]: agree, will remove

File: .vscode/settings.json
Comment (JonathanYK) [reply .vscode/settings.json:1]: agree, will remove

File: pulu/awsSecrets.json (hunk 1 above)
Comment (JonathanYK) [reply pulu/awsSecrets.json]: not possible to add comments to json file :/ so I'll just remove the values

File: pulu/initUsersTable.py (hunk 8 above)
Comment (JonathanYK) [reply pulu/initUsersTable.py]: agree, I'll add it

File: pulu/retriveAwsSecrets.js
Comment (JonathanYK) [reply pulu/retriveAwsSecrets.js:1]: I'll add the explanation to pulu/retriveAwsSecrets.ts file as this .js file is generated from this file.

File: pulu/retriveAwsSecrets.js (hunk 11 above)
Comment (JonathanYK) [reply pulu/retriveAwsSecrets.js:74]: auto-generated code from pulu/retriveAwsSecrets.ts

File: serv/.env (hunk 12 above)
Comment (JonathanYK) [reply serv/.env:4]: as a reminder, I really prefer to leave it as a comment

File: serv/Dockerfile (hunk 13 above)
Comment (JonathanYK) [reply serv/Dockerfile:1]: to add an alias for this container image (for readability and description)

File: serv/docker-compose.yml (hunk 14 above)
Comment (JonathanYK) [reply serv/docker-compose.yml]: right, will do

File: serv/src/db/init.sql (hunk 15 above)
Comment (JonathanYK) [reply serv/src/db/init.sql:3]: have to add privileges to 'compose' user to create a table, I can reduce the privileges to only needed and not all of them.

File: serv/src/index.ts (hunk 16 above)
Comment (JonathanYK) [reply serv/src/index.ts:10]: there is a validation on all mandatory params below, isn't that enough?

File: serv/tsconfig.json
Comment (JonathanYK) [reply serv/tsconfig.json:1]: I think it is related to the previous comment, as it validates the mandatory params (reqEnvVars)

File: pulu/retriveAwsSecrets.ts
Comment (JonathanYK) [reply pulu/retriveAwsSecrets.ts:1]: I'll add

**Comment by JonathanYK** [comment]
The pipeline will fail from now on as the AWS user was deleted. To review the actions - use passed action flows.

File: README.md [hunk 17]
```diff
@@ -1 +1,58 @@
-# Candidates_Server
\ No newline at end of file
+## Candidates Server: APIs and local execution
+
+The Candidates Server is developed using Express.js framework and offers several APIs for accessing and managing candidate information stored in PostgreSQL.
+
+Exposed APIs (Postman collection json [here](https://github.com/JonathanYK/Candidates_Server/tree/part2pulu/postman_collection_API_tests)):
+```
+http://<IP>:<PORT>                                                                                         - main page with greeting
+http://<IP>:<PORT>/health                                                                                  - GET health check.
```
Comment (seniakalma) [inline README.md:8]: Can be just /health (but it's really ok)

File: README.md [hunk 18]
```diff
@@ -1 +1,58 @@
...
+          aws configure set aws_access_key_id ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws configure set aws_secret_access_key ${{ secrets.AWS_ACCESS_KEY }}
+          aws configure set default.region "${{ env.AWS_REGION }}"
+
+Docker authentication with ECR:
+          aws ecr get-login-password --region ${{ env.AWS_REGION }} | docker login --username ${{ secrets.AWS_ACCOUNT_USERNAME }} --password-stdin ${{ secrets.AWS_ACCOUNT_ID }}.dkr.ecr.${{ env.AWS_REGION }}.amazonaws.com/${{ env.DEV_REPO_NAME }}
+          
+```
+**Deploying both stacks using the command `pulumi up -s <stack name>` after installing Pulumi, configuring AWS credentials, and authenticating Docker with ECR**
+- Any push or pull request to branches other than `main` will trigger [pulumi_dev_deployment action](https://github.com/JonathanYK/Candidates_Server/blob/part2pulu/.github/workflows/pulumi_dev_deployment.yml)
```
Comment (seniakalma) [inline README.md:49]: `dev` branch you meant probably

File: README.md [hunk 19]
```diff
@@ -1 +1,58 @@
...
+- While any push or pull request directly to `main` branch will trigger [pulumi_prod_deployment action](https://github.com/JonathanYK/Candidates_Server/blob/part2pulu/.github/workflows/pulumi_prod_deployment.yml)
+
+$~~~$
+
+## Infrastructure Visualization of `dev` or `prod` stacks
+
+![AWS_Infra_Visualization](https://github.com/JonathanYK/Candidates_Server/blob/part2pulu/pulu/AWS_Infra_Visualization.png?raw=true)
```
Comment (seniakalma) [inline README.md:56]: Awesome!!

File: pulu/initUsersTable.py [hunk 20]
```diff
@@ -0,0 +1,65 @@
+"""
+This Python script connects to a PostgreSQL database and creates a table named "users".
+It takes arguments for the database: host, port, name, username, and password.
```
Comment (seniakalma) [inline pulu/initUsersTable.py:3]: Awesome!!

File: pulu/retriveAwsSecrets.ts [hunk 21]
```diff
@@ -0,0 +1,53 @@
+/*
+This file uses the AWS Secrets Manager to retrieve a secret named "pulumi-rds-secrets" (template method provided by AWS)
+and saves it to a local file named "awsSecrets.json".
```
Comment (seniakalma) [inline pulu/retriveAwsSecrets.ts:3]: "This file will be later used by Pulumi to set the secrets."

File: README.md (hunk 18 above)
Comment (JonathanYK) [reply README.md:49]: dev branch, or any branch that not the main branch (according to the requirements of this task)

File: README.md (hunk 17 above)
Comment (JonathanYK) [reply README.md:8]: right, but it's more comfortable for me (LOL) to see the full url

File: pulu/dev/index.ts (hunk 5 above)
Comment (JonathanYK) [reply pulu/dev/index.ts:383]: great, added!

File: pulu/dev/index.ts (hunk 3 above)
Comment (JonathanYK) [reply pulu/dev/index.ts:18]: a dedicated file that holds the secrets, I'll update the comment

File: pulu/dev/index.ts (hunk 2 above)
Comment (JonathanYK) [reply pulu/dev/index.ts:10]: done!

**Comment by JonathanYK** [comment]
all changes were added to branch [PR-fixes](https://github.com/JonathanYK/Candidates_Server/tree/PR-fixes)

**Pull request merged**